src/rtprecv.c

```diff
@@ -0,0 +1,739 @@
+			     src, hdr->seq, hdr->ts, err);
+			metric_inc_err(rx->metric);
+		}
+
+		uint32_t n = jbuf_packets(rx->jbuf);
+		while (n--) {
+			if (decode_frame(rx) != EAGAIN)
+				break;
+		}
+	}
+	else {
+		(void)handle_rtp(rx, hdr, mb, 0, false);
+	}
+}
+
+
+void rtprecv_handle_rtcp(const struct sa *src, struct rtcp_msg *msg,
+			  void *arg)
+{
+	struct rtp_receiver *rx = arg;
+	(void)src;
+
+	MAGIC_CHECK(rx);
+
+	mtx_lock(rx->mtx);
+	rx->ts_last = tmr_jiffies();
+	mtx_unlock(rx->mtx);
+
+	pass_rtcp_work(rx, msg);
+}
+
+
+void rtprecv_mnat_connected_handler(const struct sa *raddr1,
+				     const struct sa *raddr2, void *arg)
+{
+	struct rtp_receiver *rx = arg;
+
+	MAGIC_CHECK(rx);
+
+	pass_mnat_work(rx, raddr1, raddr2);
+}
+
+
+/*
+ * functions that run in main thread
+ */
+
+void rtprecv_set_ssrc(struct rtp_receiver *rx, uint32_t ssrc)
+{
+	mtx_lock(rx->mtx);
+	if (rx->ssrc_set) {
+		if (ssrc != rx->ssrc) {
+			debug("stream: receive: SSRC changed: %x -> %x\n",
+			     rx->ssrc, ssrc);
+			rx->ssrc = ssrc;
+		}
+	}
+	else {
+		debug("stream: receive: setting SSRC: %x\n", ssrc);
+		rx->ssrc = ssrc;
+		rx->ssrc_set = true;
+	}
+	mtx_unlock(rx->mtx);
+}
+
+
+uint64_t rtprecv_ts_last(struct rtp_receiver *rx)
+{
+	uint64_t ts_last;
+	mtx_lock(rx->mtx);
+	ts_last = rx->ts_last;
+	mtx_unlock(rx->mtx);
+
+	return ts_last;
+}
+
+
+void rtprecv_set_ts_last(struct rtp_receiver *rx, uint64_t ts_last)
+{
+	mtx_lock(rx->mtx);
+	rx->ts_last = ts_last;
+	mtx_unlock(rx->mtx);
+}
+
+
+void rtprecv_flush(struct rtp_receiver *rx)
+{
+	if (!rx)
+		return;
+
+	jbuf_flush(rx->jbuf);
+}
+
+
+void rtprecv_set_enable(struct rtp_receiver *rx, bool enable)
+{
+	if (!rx)
+		return;
+
+	mtx_lock(rx->mtx);
+	rx->enabled = enable;
+	mtx_unlock(rx->mtx);
+}
+
+
+int rtprecv_get_ssrc(struct rtp_receiver *rx, uint32_t *ssrc)
+{
+	int err;
+
+	if (!rx || !ssrc)
+		return EINVAL;
+
+	mtx_lock(rx->mtx);
+	if (rx->ssrc_set) {
+		*ssrc = rx->ssrc;
+		err = 0;
+	}
+	else
+		err = ENOENT;
+	mtx_unlock(rx->mtx);
+
+	return err;
+}
+
+
+void rtprecv_enable_mux(struct rtp_receiver *rx, bool enable)
+{
+	mtx_lock(rx->mtx);
+	rtcp_enable_mux(rx->rtp, enable);
+	mtx_unlock(rx->mtx);
+}
+
+
+/**
+ * The debug function prints into the given mbuf in order to avoid long
+ * blocking print to stdout.
+ *
+ * @param rx The rtp_receiver
+ * @param mb Memory buffer
+ *
+ * @return 0 if success, otherwise errorcode
+ */
+int rtprecv_debug(struct re_printf *pf, const struct rtp_receiver *rx)
+{
+	int err;
+	bool enabled;
+
+	mtx_lock(rx->mtx);
+	enabled = rx->enabled;
+	mtx_unlock(rx->mtx);
+
+	err  = re_hprintf(pf, " rx.enabled: %s\n", enabled ? "yes" : "no");
+	err |= jbuf_debug(pf, rx->jbuf);
+
+	return err;
+}
+
+
+static void destructor(void *arg)
+{
+	struct rtp_receiver *rx = arg;
+	bool join = false;
+
+	mtx_lock(rx->mtx);
+	if (re_atomic_rlx(&rx->run)) {
+		join = true;
+		re_atomic_rlx_set(&rx->run, false);
+	}
+	mtx_unlock(rx->mtx);
+	if (join)
+		thrd_join(rx->thr, NULL);
```

Comment:
I think this can be simplified now (do not commit the suggestion, only used for nice diff):

```suggestion
	if (re_atomic_rlx(&rx->run)) {
		re_atomic_rlx_set(&rx->run, false);
		thrd_join(rx->thr, NULL);
	}
```

Reply:
I think if the thread terminates before `thrd_join()` then we run into troubles. We had such problems already, right?

Reply:
A exited thread is fine (man pthread_join):
```
The  pthread_join() function waits for the thread specified by thread to terminate.  
If that thread has already terminated, then pthread_join() returns immediately. 
```
Only two concurrent joins are problematic.

Reply:
It's only problmatic if `thrd_join()` is missing and not called.

Reply:
ok, thanks!